### PR TITLE
feat: Support Timestamp +/- Interval types

### LINKED ISSUE #3963: Support  `Timestamp` `+`/`-` `Interval` types
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
In DataFusion we would like to add and subtract timestamps. While we currently support this for scalar (singular) values we don't support it for arrays. 

So this query will work

```sql
select * from foo where now() - INTERVAL '1 day' 
```

This one will not (aka add `time` column to a `Interval` column): 

```sql
select * from foo where now() > foo.time + foo.sale_duration
```

**Describe the solution you'd like**
I would like the `add_dyn` and `subtract_dyn` kernels to handle

`Timestamp(Second, ..)` + `Interval(YearMonth)`
`Timestamp(Second, ..)` + `Interval(DayTime)`
`Timestamp(Second, ..)` + `Interval(MonthDayNano)`

`Timestamp(Millisecond, ..)` + `Interval(YearMonth)`
`Timestamp(Millisecond, ..)` + `Interval(DayTime)`
`Timestamp(Millisecond, ..)` + `Interval(MonthDayNano)`

`Timestamp(MicroSecond, ..)` + `Interval(YearMonth)`
`Timestamp(Microsecond, ..)` + `Interval(DayTime)`
`Timestamp(Microsecond, ..)` + `Interval(MonthDayNano)`

`Timestamp(Nanosecond, ..)` + `Interval(YearMonth)`
`Timestamp(Nanosecond, ..)` + `Interval(DayTime)`
`Timestamp(Nanosecond, ..)` + `Interval(MonthDayNano)`

To keep consistency with https://github.com/apache/arrow-rs/issues/3959 and other timestamp kernels:
1. The precision will be truncated if necessary (e.g. adding an `IntervalMonthDayNano` with nanosecond precision to a `Timestamp(Millisecond) will not error and the interval will be truncated to  millisecond precision)
2. If an overflow occurs (e.g. an `Interval(DayTime)` is added to `Timestamp(Nanosecond)` that goes beyond the range of i64 in nanoseconds), an error would be raised

**Describe alternatives you've considered**

**Additional context**
Here is the support for `Date + Interval` -- we should be able to follow much the same pattern for `Timestamp + interval` and `Timestamp - Interval`: 

https://github.com/apache/arrow-rs/blob/9bd2bae586ed5b0edfd699f89a0855d79f61b611/arrow-arith/src/arithmetic.rs#L694-L740

Here is example code showing what the desire is;

```rust
fn add_timestamp_intervals() {
    // Given some timestamps
    let arr = TimestampNanosecondArray::from(vec![
        1_000,
        1_000_000,
        1_000_000_000,
        1_000_000_000_000,
        1_000_000_000_000_000,
        1_000_000_000_000_000_000,
        2_000_000_000_000_000_000,
    ]);

    let types = vec![
        DataType::Timestamp(TimeUnit::Second, None),
        DataType::Timestamp(TimeUnit::Millisecond, None),
        DataType::Timestamp(TimeUnit::Microsecond, None),
        DataType::Timestamp(TimeUnit::Nanosecond, None),
    ];

    let input_date_time_arrays: Vec<_> = types
        .iter()
        .map(|dt| cast(&arr, dt).unwrap())
        .collect();

    for date_time_array in input_date_time_arrays.into_iter() {


        // would like to add 1 day
        let interval_array = IntervalDayTimeArray::from(vec![
            Some(IntervalDayTimeType::make_value(1, 0)),
            Some(IntervalDayTimeType::make_value(1, 0)),
            Some(IntervalDayTimeType::make_value(1, 0)),
            Some(IntervalDayTimeType::make_value(1, 0)),
            Some(IntervalDayTimeType::make_value(1, 0)),
            Some(IntervalDayTimeType::make_value(1, 0)),
            Some(IntervalDayTimeType::make_value(1, 0)),
        ]);


        // input Timestamp(Second, None):
        // +---------------------+
        // | input               |
        // +---------------------+
        // | 1970-01-01T00:00:00 |
        // | 1970-01-01T00:00:00 |
        // | 1970-01-01T00:00:01 |
        // | 1970-01-01T00:16:40 |
        // | 1970-01-12T13:46:40 |
        // | 2001-09-09T01:46:40 |
        // | 2033-05-18T03:33:20 |
        // +---------------------+
        println!(
            "input {}:\n{}",
            date_time_array.data_type(),
            pretty_format_columns("input", &[Arc::new(date_time_array.clone()) as _]).unwrap()
        );



        let col_name = format!(
            "{} + {}",
            date_time_array.data_type(),
            interval_array.data_type()
        );

        // thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: CastError("Unsupported data type Timestamp(Second, None), Interval(DayTime)")', src/main.rs:253:66
        let new_arr = add_dyn(&date_time_array, &interval_array).unwrap();
        println!(
            "{col_name}:\n{}",
            pretty_format_columns("add", &[new_arr]).unwrap()
        );

        let col_name = format!(
            "{} - {}",
            date_time_array.data_type(),
            interval_array.data_type()
        );

        let new_arr = subtract_dyn(&date_time_array, &interval_array).unwrap();
        println!(
            "{col_name}:\n{}",
            pretty_format_columns("sub", &[new_arr]).unwrap()
        );

    }

}
```

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #3963

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/types.rs [hunk 1]
```diff
@@ -350,6 +350,650 @@ impl ArrowTimestampType for TimestampNanosecondType {
...
+        let res = res
+            .checked_add_signed(Duration::milliseconds(ms as i64))
+            .ok_or_else(|| {
+                ArrowError::ComputeError("Timestamp out of range".to_string())
+            })?;
+        TimestampMillisecondType::make_value(res)
+            .ok_or_else(|| ArrowError::ComputeError("Timestamp out of range".to_string()))
+    }
+
+    /// Adds the given IntervalMonthDayNanoType to an arrow TimestaTimestampMillisecondTypempSecondType
```
Comment (metesynnada) [inline arrow-array/src/types.rs:718]:
```suggestion
    /// Adds the given IntervalMonthDayNanoType to an arrow TimestampMillisecondType
```

File: arrow-arith/src/arithmetic.rs [hunk 2]
```diff
@@ -739,6 +739,105 @@ pub fn add_dyn(left: &dyn Array, right: &dyn Array) -> Result<ArrayRef, ArrowErr
                 ))),
             }
         }
+        DataType::Timestamp(TimeUnit::Second, _) => {
+            let l = left.as_primitive::<TimestampSecondType>();
+            match right.data_type() {
+                DataType::Interval(IntervalUnit::YearMonth) => {
+                    let r = right.as_primitive::<IntervalYearMonthType>();
+                    let res = math_checked_op(l, r, TimestampSecondType::add_year_months)?;
+                    Ok(Arc::new(res))
```
Comment (metesynnada) [inline arrow-arith/src/arithmetic.rs:748]: You can use `Arc::new(res.with_timezone_opt(TZ))` to support timezone output (in all match arms) where `TZ` in `DataType::Timestamp(TimeUnit::Second, TZ)`

**Review by metesynnada — COMMENTED** [review]
PR looks neat! The PR performs arithmetic with respect to the UTC epoch, and it lacks timezone support on output. You can utilize `with_timezone_opt()` on output to support timezone information.

Maybe you can check the discussion under https://github.com/apache/arrow-datafusion/pull/5764.

**Comment by metesynnada** [comment]
LGTM!

**Review by alamb — COMMENTED** [review]
Thank you @Weijun-H  -- this looks epic. 

I think this PR needs a little more test coverage (ensure that all fields are non zero for some tests) and there is an outstanding comment by @metesynnada but otherwise this looks good to me

cc @tustvold  in case you would like to weigh in

**Comment by alamb** [comment]
Also, when testing locally I got this error I think because this PR uses APIs that were introduced in chrono `0.4.24`

```
   Compiling arrow-array v37.0.0 (/Users/alamb/Software/arrow-rs2/arrow-array)
error[E0599]: no function or associated item named `from_timestamp_micros` found for struct `NaiveDateTime` in the current scope
   --> arrow-array/src/types.rs:520:36
    |
520 |         let prior = NaiveDateTime::from_timestamp_micros(timestamp).ok_or_else(|| {
    |                                    ^^^^^^^^^^^^^^^^^^^^^
    |                                    |
    |                                    function or associated item not found in `NaiveDateTime`
    |                                    help: there is an associated function with a similar name: `from_timestamp`


```

I needed to make this change (which I think we should do in the PR):


```diff
diff --git a/arrow-array/Cargo.toml b/arrow-array/Cargo.toml
index 1b417bb0e..634a0aa64 100644
--- a/arrow-array/Cargo.toml
+++ b/arrow-array/Cargo.toml
@@ -44,7 +44,7 @@ ahash = { version = "0.8", default-features = false, features = ["runtime-rng"]
 arrow-buffer = { workspace = true }
 arrow-schema = { workspace = true }
 arrow-data = { workspace = true }
-chrono = { version = "0.4.23", default-features = false, features = ["clock"] }
+chrono = { version = "0.4.24", default-features = false, features = ["clock"] }
 chrono-tz = { version = "0.8", optional = true }
 num = { version = "0.4", default-features = false, features = ["std"] }
 half = { version = "2.1", default-features = false, features = ["num-traits"] }
```

File: arrow-arith/src/arithmetic.rs [hunk 3]
```diff
@@ -3665,4 +3863,540 @@ mod tests {
             "1234567890.0000000000000000000000000000"
         );
     }
+
+    #[test]
+    fn test_timestamp_second_add_interval() {
+        // timestamp second + interval year month
+        let a = TimestampSecondArray::from(vec![1, 2, 3, 4, 5]);
+        let b = IntervalYearMonthArray::from(vec![
+            Some(IntervalYearMonthType::make_value(1, 1)),
```
Comment (alamb) [inline arrow-arith/src/arithmetic.rs:3872]: The reason I was suggesting tests that had different non zero fields (e.g. `years` and `month`s in this example) 

e.g.
```rust 
            Some(IntervalYearMonthType::make_value(1, 2)),
```

Would be to catch errors such as when the year and month fields were mixed up in the code. For example, with

```rust
            Some(IntervalYearMonthType::make_value(1, 1)),
```

If the code made a mistake with which field , these tests would still pass.

**Review by alamb — APPROVED** [review]
Thank you @Weijun-H  -- this PR looks good to me. Thank you to @metesynnada  for the review.

@tustvold @waitingkuo  or @avantgardnerio  would anyone else like to review this PR prior to merge?

File: arrow-arith/src/arithmetic.rs (hunk 3 above)
Comment (alamb) [reply arrow-arith/src/arithmetic.rs:3872]: While I still think the tests can be improved, I also think they are good enough to merge

**Comment by alamb** [comment]
Here is a PR to bump the chrono version:  https://github.com/apache/arrow-rs/pull/4093

File: arrow-arith/src/arithmetic.rs [hunk 4]
```diff
@@ -739,6 +739,105 @@ pub fn add_dyn(left: &dyn Array, right: &dyn Array) -> Result<ArrayRef, ArrowErr
                 ))),
             }
         }
+        DataType::Timestamp(TimeUnit::Second, _) => {
```
[the remark this answers is not in the document]
Comment (alamb) [reply arrow-arith/src/arithmetic.rs:742]: Follow on ticket: https://github.com/apache/arrow-rs/issues/4094

**Review by alamb — APPROVED** [review]
Thanks @Weijun-H

**Pull request merged**